llvm/unittests/SandboxIR/PassTest.cpp

```diff
@@ -86,6 +87,68 @@ define void @foo() {
 #endif
 }
 
+TEST_F(PassTest, RegionPass) {
+  auto *F = parseFunction(R"IR(
+define i8 @foo(i8 %v0, i8 %v1) {
+  %t0 = add i8 %v0, 1
+  %t1 = add i8 %t0, %v1, !sandboxvec !0
+  %t2 = add i8 %t1, %v1, !sandboxvec !0
+  ret i8 %t1
+}
+
+!0 = distinct !{!"sandboxregion"}
+)IR",
+                          "foo");
+
+  class TestPass final : public RegionPass {
+    unsigned &InstCount;
+
+  public:
+    TestPass(unsigned &InstCount)
+        : RegionPass("test-pass"), InstCount(InstCount) {}
+    bool runOnRegion(Region &R) final {
+      for ([[maybe_unused]] auto &Inst : R) {
+        ++InstCount;
+      }
+      return false;
+    }
+  };
+  unsigned InstCount = 0;
+  TestPass TPass(InstCount);
+  // Check getName(),
+  EXPECT_EQ(TPass.getName(), "test-pass");
+  // Check runOnRegion();
+  llvm::SmallVector<std::unique_ptr<Region>> Regions =
+      Region::createRegionsFromMD(*F);
+  ASSERT_EQ(Regions.size(), 1u);
+  TPass.runOnRegion(*Regions[0]);
+  EXPECT_EQ(InstCount, 2u);
+#ifndef NDEBUG
+  {
+    // Check print().
+    std::string Buff;
+    llvm::raw_string_ostream SS(Buff);
+    TPass.print(SS);
+    EXPECT_EQ(Buff, "test-pass");
+  }
+  {
+    // Check operator<<().
+    std::string Buff;
+    llvm::raw_string_ostream SS(Buff);
+    SS << TPass;
+    EXPECT_EQ(Buff, "test-pass");
+  }
+  // Check pass name assertions.
+  class TestNamePass final : public RegionPass {
+  public:
+    TestNamePass(llvm::StringRef Name) : RegionPass(Name) {}
+    bool runOnRegion(Region &F) { return false; }
+  };
+  EXPECT_DEATH(TestNamePass("white space"), ".*whitespace.*");
```

Comment:
Do we have a comment at the constructor about these name restrictions?

Reply:
The base `Pass` class does, kinda (it mentions it can't contain spaces, but it doesn't mention it can't begin with a `-`). And the comment is not for the constructor but for the `Name` member. So, technically the answer to your question is "no, not really, and neither does `FunctionPass`".

Reply:
Maybe we should have some kind of comment that describes the allowed names or points to the code that does the checks.

Reply:
Would something like this suffice?

Reply:
Yeah, looks great.

Reply:
I feel like the asserts are pretty self-explanatory and easy to fix if someone hits them. Maybe adding the reasoning behind those restrictions in comments around the asserts is good

Reply:
The `Name` field in the `Pass` class already has a comment with the reasoning.

```
/// The pass name. This is also used as a command-line flag and should not
/// contain whitespaces.
```

Reply:
ah ok

unrelated to this patch, I'm not sure why we don't allow "-" as the first char. it's not conventional to start a pass name with "-", but there are plenty of other worse things you could do, like start with "("

Reply:
Yes, using an illegal name shouldn't be too much of an issue. Feel free to drop the comments you added.

Reply:
I'll merge as-is if you don't mind. The comments are a slight improvement IMO (nothing wrong to document the public interface of the class) and I don't want to push another revision and start pre-merge checks again.